### ISSUE TITLE
nvim_win_set_option should not set global value for window-local options

### ISSUE THREAD

**Issue by ddickstein** [neovim/neovim#20045]
### Neovim version (nvim -v)

0.7.0

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

CentOS 7.9

### Terminal name/version

xfce4-terminal 0.8.7.4

### $TERM environment variable

tmux-256color

### Installation

system package manager

### How to reproduce the issue

`nvim_win_set_option` sets the global value for window-local options. It looks like this was [intentional](https://github.com/neovim/neovim/pull/9110), but I think it is a mistake. The global value and the local value of a window-local option are meaningfully distinct, and at different times I may want to set one without setting the other. While this can be done using `nvim_set_option_value` and passing the appropriate `scope` argument, that only works for the current window.

I'm having trouble understanding the rationale for changing both the local and global value together. The closest I've found is [this explanation](https://github.com/neovim/neovim/issues/15529#issuecomment-908814854):

> It looks like this is intentional (see https://github.com/neovim/neovim/pull/9110). The semantics of :setlocal and vim.opt_local (which uses nvim_win_set_option) are different with window options. For example, even though 'number' claims to be a window-local option, if you :setlocal number and then open a new buffer in the same window, 'number' will be disabled. So I think vim.opt_local is actually doing the right thing and :setlocal is in the wrong in this case.

This is wrong. `number` is indeed a window-local option, and if you `:setlocal number` and then open a new buffer in the same window, `number` is disabled *correctly*, as this is how window-local options are expected to behave. For a brand new buffer, the local value for the window-local option is initialized to the window's global value for the option, which is not modified by `setlocal`. If the user wanted to set the window's global value for the window-local option they could have used `set number`.

[Here](https://github.com/neovim/neovim/issues/19549)'s another report of the same issue, and the current behavior of `nvim_win_set_option` is stated but isn't justified.

Can this decision be revisited? The [proposal](https://github.com/neovim/neovim/pull/9110#issue-369496905) to have `nvim_buf_set_option` work for the "local" value of window-local settings makes sense on some level (the override is local to the buffer in the window) but is unintuitive from an API standpoint. I think a scope argument should be added to `nvim_win_set_option`. This isn't needed for `nvim_buf_set_option` because `nvim_set_option` (or `nvim_set_option_value` with global scope) can be used to set the global value. However, this is insufficient for window-local options because the global value is per-window and these functions only operate in the current window.

Note that for "global or local to window" options `nvim_win_set_option` just sets the local value, which is probably more "correct" than setting both the local and global values, but the API is similarly limited - there isn't a way to set the global value for a different window.

### Expected behavior

There should be a way to independently set the global and local values for window-local options.

### Actual behavior

There is no way to independently set the global and local values for window-local options.

**Comment by lewis6991** [comment]
> nvim_win_set_option sets the global value for window-local options.

Window local options do not have a global value, except for global-local windows options (see `:help global-local`). `set winopt` does not set a global value, it sets a value for all buffers in a specific window. `setlocal winopt` sets a value for a specific buffer in a window.

> While this can be done using nvim_set_option_value and passing the appropriate scope argument, that only works for the current window.

RTFM

```vimhelp
                                                     *nvim_set_option_value()*
nvim_set_option_value({name}, {value}, {*opts})
    Sets the value of an option. The behavior of this function matches that of
    |:set|: for global-local options, both the global and local value are set
    unless otherwise specified with {scope}.

    Note the options {win} and {buf} cannot be used together.

    Parameters: ~
        {name}   Option name
        {value}  New option value
        {opts}   Optional parameters
                 • scope: One of 'global' or 'local'. Analogous to
                   |:setglobal| and |:setlocal|, respectively.
                 • win: |window-ID|. Used for setting window local option.
                 • buf: Buffer number. Used for setting buffer local option.
```

> https://github.com/neovim/neovim/issues/19549's another report of the same issue, and the current behavior of nvim_win_set_option is stated but isn't justified.

It does not need to be justified. It was designed to work the way it does and due to the API contract that cannot be changed. This is why `nvim_set_option_value` was introduced and effectively deprecates `nvim_win_set_option`.

> There should be a way to independently set the global and local values for window-local options.

Use `nvim_set_option_value`

**Comment by ddickstein** [comment]
> Window local options do not have a global value, except for global-local windows options (see :help global-local). set winopt does not set a global value, it sets a value for all buffers in a specific window. setlocal winopt sets a value for a specific buffer in a window.

I understand these semantics. When I used the term "global value" for window-local options, I meant it to describe the value that is per-window but global to all buffers in the window. This is how `:h local-options` suggests using the term "global" for window-local options:

> Vim keeps a global value of the local window options, which is used when editing another buffer.  Each window
has its own copy of these values.  Thus these are local to the window, but global to all buffers in the window.

---

> RTFM

With all due respect, I did read the manual - the relevant part wasn't included in it. The `win` and `buf` options were added to `nvim_set_option_value` in https://github.com/neovim/neovim/commit/6219331c4d333e5a76129746021f972c21a040db, which was after the 0.7.0 release. I didn't know this had been addressed since then.

> It does not need to be justified.

I didn't mean to suggest that justification is owed, only to observe that justification wasn't provided, which makes it harder to understand the rationale behind the design choice. As far as I can tell, the change to `nvim_set_option_value` to support `win` and `buf` and effectively deprecate these functions moots the issue, but before this change I think it reflected a material limitation in the API.

**Comment by bfredl** [comment]
> https://github.com/neovim/neovim/issues/19549's another report of the same issue, and the current behavior of nvim_win_set_option is stated but isn't justified.

@ddickstein  Both possible behaviors are _useful_ and thus _justified_. If we had chosen the other behavior, i e `setlocal`, another user would have spilled the same amount of ink as you just did about how its **obviously wrong** and **not justified** that nvim_set_win_option would not set the value for the window but just for the buffer. They would likewise quote a bunch of random issues as "proof" as our current behavior is somehow broken and their personal preference the correct one.

**Comment by ddickstein** [comment]
I completely agree with you that both possible behaviors are useful, and that either setting just the local value or setting both the local and global value would leave a use case unaddressed. For that reason, either of those behaviors would be broken / incorrect, irrespective of personal preference / intuition. That's why when I created the issue I didn't propose changing the behavior to just affect the local value, I proposed adding a scope argument that would allow you to specify which value you wanted to affect. @lewis6991 pointed out that I wasn't considering API stability in that proposal, and I think the change to `nvim_set_option_value` to include `win` and `buf` alongside a `scope` argument meets the behavior I was seeking.

I think my word choice created some wrong impressions here. I neither meant to suggest that a justification is "owed" nor meant to suggest that these issues are somehow "proof" that the behavior of `nvim_win_set_option` is broken. I meant to observe that there is a limitation in `nvim_win_set_option`, and that after reviewing reported issues about the function (to show that I looked before reporting an issue), I could not find an explanation of why this limitation was present. By "I think it is a mistake," I meant that coupling global & local together as opposed to splitting them apart is a mistake, not that setting both as opposed to just setting local is a mistake.
